File: reactive-sql-client-examples/src/main/js/io/vertx/example/sqlclient/simple/sql_client_example.js
var PgPool = require("vertx-pg-client-js/pg_pool");

var pool = PgPool.pool(vertx, {
  "port" : 5432,
  "host" : "the-host",
  "database" : "the-db",
  "user" : "user",
  "password" : "secret"
}, {
  "maxSize" : 4
});

// Uncomment for MySQL
//    Pool pool = MySQLPool.pool(vertx, new MySQLConnectOptions()
//      .setPort(5432)
//      .setHost("the-host")
//      .setDatabase("the-db")
//      .setUser("user")
//      .setPassword("secret"), new PoolOptions().setMaxSize(4));

pool.getConnection(function (res1, res1_err) {
  if (res1_err != null) {
    console.error(res1_err.getMessage());
    return
  }
  var connection = res1;

  // create a test table
  connection.query("create table test(id int primary key, name varchar(255))").execute(function (res2, res2_err) {
    if (res2_err != null) {
      connection.close();
      console.error("Cannot create the table");
      res2_err.printStackTrace();
      return
    }

    // insert some test data
    connection.query("insert into test values (1, 'Hello'), (2, 'World')").execute(function (res3, res3_err) {

      // query some data with arguments
      connection.query("select * from test").execute(function (rs, rs_err) {
        if (rs_err != null) {
          console.error("Cannot retrieve the data from the database");
          rs_err.printStackTrace();
          return
        }

        Array.prototype.forEach.call(rs, function(line) {
          console.log("" + line);
        });

        // and close the connection
        connection.close();
      });
    });
  });
});

File: sql-client-examples/src/main/java/io/vertx/example/sqlclient/streaming/SqlClientExample.java
package io.vertx.example.sqlclient.streaming;

import io.vertx.core.*;
import io.vertx.pgclient.PgConnectOptions;
import io.vertx.sqlclient.*;
import org.testcontainers.containers.PostgreSQLContainer;

/*
 * @author <a href="mailto:pmlopes@gmail.com">Paulo Lopes</a>
 */
public class SqlClientExample extends VerticleBase {

  // Convenience method so you can run it in your IDE
  public static void main(String[] args) {
    PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>();
    postgres.start();
    PgConnectOptions options = new PgConnectOptions()
      .setPort(postgres.getMappedPort(5432))
      .setHost(postgres.getContainerIpAddress())
      .setDatabase(postgres.getDatabaseName())
      .setUser(postgres.getUsername())
      .setPassword(postgres.getPassword());
    // Uncomment for MySQL
//    MySQLContainer<?> mysql = new MySQLContainer<>();
//    mysql.start();
//    MySQLConnectOptions options = new MySQLConnectOptions()
//      .setPort(mysql.getMappedPort(3306))
//      .setHost(mysql.getContainerIpAddress())
//      .setDatabase(mysql.getDatabaseName())
//      .setUser(mysql.getUsername())
//      .setPassword(mysql.getPassword());
    Vertx vertx = Vertx.vertx();
    vertx.deployVerticle(new SqlClientExample(options));  }

  private final SqlConnectOptions options;
  private Pool pool;

  public SqlClientExample(SqlConnectOptions options) {
    this.options = options;
  }

  @Override
  public Future<?> start() {

    pool = Pool.pool(vertx, options, new PoolOptions().setMaxSize(4));

    return pool.withConnection(connection -> {
      // create a test table
      return connection
        .query("create table test(id int primary key, name varchar(255))")
        .execute()
        .compose(v -> {
          // insert some test data
          return connection
            .query("insert into test values (1, 'Hello'), (2, 'World')")
            .execute();
        })
        .compose(v ->  connection
          .prepare("select * from test")
          .compose(ps -> {
            RowStream<Row> stream = ps.createStream(50);
            Promise<Void> promise = Promise.promise();
            stream
              .exceptionHandler(promise::fail)
              .endHandler(promise::complete)
              .handler(row -> System.out.println("row = " + row.toJson()));
            return promise
              .future()
              .eventually(ps::close);
          }));
    }).onSuccess(ar -> {
      System.out.println("done");
    });
  }
}
